### DISCUSSION TITLE
Mitigate enterprise anti-malware problems

### DISCUSSION THREAD

**Discussion by gnachman** [sparkle-project/Sparkle#2641] [Ideas]
## Summary

Every time I update my app, I get many responses from users whose upgrades failed because SentinelOne quarantined the new version. They have to beg their administrators to add the app to an allow-list and try again. It would be nice if there were a mechanism to detect this problem and educate the users about it instead of leaving them with a broken installation.

## Possible Fix

After unzipping the app, SentinelOne will delete the main app binary. I propose that an app which uses sparkle be given the ability to opt into an extra verification step.  After unzipping the payload, wait for a configurable delay and then check that it is intact (could check code signature or verify that the files in the app bundle match the files in the zip). If it detects a problem, show a message like "It looks like the app update was quarantined by your anti-malware. Please contact your system administrator".

**Comment by zorgiepoo** [comment]
[I've been bitten](https://github.com/zorgiepoo/Bit-Slicer/issues/70) by SentinelOne as well, although through a different issue.

I don't like the idea of adding a delayed/periodic code signing check in the installer tool (but if one were to be added I think it'd be automatic and not configurable -- and I think the message ought to be reworded; it could also be malware doing this, or something else going catastrophically wrong).

I am also not confident an additional check will help much, if such software can/will damage the app after Sparkle has installed the new app in /Applications, or wherever the old app was originally, instead of in the staging (Caches) area, which can be a race. (And there may be a lack of proof it's damaging the app before Sparkle moves it versus after the move. I understand you are getting reports from the app being damaged after the user manually unzips a downloaded app but that may not be the same circumstance).

There are some potential app tampering protections but they are kind of moot as software like this is typically given full access / permissions to the system.

**Reply by gnachman** [reply]
What if it were temporarily installed in a location under the original app bundle? So /Applications/MyApp.app/Temp/MyApp.app. I have a feeling this is a bad idea but I can't quite put my finger on it. I don't believe code signing would be a problem as long as the app has been launched at least once.

**Reply by zorgiepoo** [reply]
It is a bad idea because it invalidates the code signature of the bundle (the system can do periodic scans too). I'm also not sure about the motivation behind that. If SentinelOne looks in Caches (which is not proved), it can just as equally well look inside other app bundles. At the end of the day, if Sparkle installs the update and the update was not corrupt at the point before the move/swap, SentinelOne can just remove it after the move/swap or later. Sparkle is not designed to protect against anti-virus software, and neither is the OS unfortunately since it's not malware classified and intentionally given full disk access by some administrator to do this type of task.

**Reply by gnachman** [reply]
To be clear, my goal is to get SentinelOne to delete the executable while we still have the old version around. Extracting it into a temporary location that we know it looks at could bait it into revealing that it considers the new version to be malware. If that happens, then I can tell the user "Don't upgrade yet! It'll break. Go beg your sys admin for an exception." and then we don't complete the upgrade.

I don't know if SentinelOne scans the caches folder you use. If it does, then introducing a delay would be sufficient. If not, we could find another spot like `~/.sparkle/`. I know that merely unzipping the executable in Downloads *is* sufficient to trigger SO's behavior.

If this is all too weird, maybe I should do it via Sparkle's delegate APIs. I'm not sure if they're capable of this or not.

**Reply by zorgiepoo** [reply]
Ah, okay didn't know that was your goal.

No telling if it'll scan ~/.sparkle either.

This is against Sparkle's normal intentions really, to more-expose the update. If anything, we'd go in the opposite direction for security against local malware.

It's not a supported path for the app to have access to the new unarchived install or anything like that. This is even more isolated in Sparkle 2.

**Reply by gnachman** [reply]
After some more investigation, I noticed that Sparkle can install a corrupt bundle (corrupted by SentinelOne) by this path:

`-[SUUpdateValidator validateUpdateForHost:downloadedToPath:newBundleURL:signatures:]` checks if the update is code signed:

```
    BOOL updateIsCodeSigned = [SUCodeSigningVerifier bundleAtURLIsCodeSigned:newHost.bundle.bundleURL];
```
If the update is code signed but its main app binary was deleted, `updateIsCodeSigned` will be `NO`. I find this very surprising. Consequently, the following code returns YES and the corrupt bundle gets installed:
```
    NSError *error = nil;
    if (passedDSACheck && updateIsCodeSigned && ![SUCodeSigningVerifier codeSignatureIsValidAtBundleURL:newHost.bundle.bundleURL error:&error]) {
        SULog(SULogLevelError, @"The update archive has a valid (Ed)DSA signature, but the app is also signed with Code Signing, which is corrupted: %@. The update will be rejected.", error);
        return NO;
    }
    // Either DSA must be valid, or Apple Code Signing must be valid.
    // We allow failure of one of them, because this allows key rotation without breaking chain of trust.
    if (passedDSACheck || passedCodeSigning) {
        return YES;
    }
```
Since the host app bundle is code signed, passing the DSA check is sufficient.

It smells to me like this is not the intended behavior and that `bundleAtURLIsCodeSigned:` is misbehaving. Unfortunately, it returns NO because `SecCodeCopyDesignatedRequirement` returns `errSecCSUnsigned`. I checked and the `_CodeSignature` directory is there. 

I don't understand all the thinking that went into the existing logic around signature checking. I believe it's pretty old and predates modern code signature requirements. I am tempted to add a clause like this:

```
    if (hostIsCodeSigned && !updateIsCodeSigned) {
        SULog(SULogLevelError, @"The original application is code signed but the update does not have a valid signature.");
        return NO;
    }
```

…but I don't know if that's actually a terrible idea :)

At any rate, if this problem is fixed then it's possible that Sparkle could already detect SentinelOne's shenanigans without having to introduce other changes.

**Reply by 5HT2** [reply]
> What if it were temporarily installed in a location under the original app bundle? So /Applications/MyApp.app/Temp/MyApp.app. I have a feeling this is a bad idea but I can't quite put my finger on it. I don't believe code signing would be a problem as long as the app has been launched at least once.

I don't know if it is a bad idea, but off the top of my head
- Jamf (RMM software) struggles to find things that aren't in the default path for software inventory checking / whitelisting[^1]
- Spotlight itself will struggle to find apps that are not literally in the root path of an applications folder, even if they are linked inside after the fact (I do think that this is a caching thing, unfortunately).
  - This appears to be fixed as of macOS 15.0, I am not sure if this was a temporary bug but it does not affect machines I have that are on ≥15.0, and the official docs imply[^2] you should be able to organize apps in folders (when this was historically taboo according to official docs).
  - Maybe this is not actually that bad of an idea as a temporary thing? It seems doable and does not cause problems on it's own, apart from "some apps don't handle updates like this very well"[^3] but it doesn't require pin entry or elevated permissions at all.

~~Otherwise, there shouldn't be an issue in that regard, especially as a temporary thing.~~

Edit: or I did not notice the reply re: code signing. This does not appear to be the case if the application has a code-signing signature that has the intention of being installed in that folder, _or_ if the file is copied over as a logged-in user:
|Original|Copy|
|---|---|
| <img width="692" alt="Screenshot 2024-10-15 at 20 17 40" src="https://github.com/user-attachments/assets/7ac073d8-95b8-49dd-bd82-cace95fa3c4f"> | <img width="692" alt="Screenshot 2024-10-15 at 20 15 43" src="https://github.com/user-attachments/assets/05b54a81-0d3a-43db-b63e-df511abc8f12"> |

[^1]: Even when given a default scan directory, such as `/Applications/NinjaRemote` which contains`/Applications/NinjaRemote/ncstreamer.app`, it will take a long time to find it / detect that it is gone or is new, especailly when machines get reimaged
[^2]: https://support.apple.com/en-ca/guide/mac-help/mh26885/mac
[^3]: https://discussions.apple.com/thread/8271392?sortBy=rank

**Reply by zorgiepoo** [reply]
> After some more investigation, I noticed that Sparkle can install a corrupt bundle (corrupted by SentinelOne) by this path:

I'm aware of this logic. If (Ed)DSA signature verification succeeds (update is trusted at that point), Sparkle does a shallow code signing check to validate basic integrity and that the developer didn't make a mistake shipping their app with an obviously malformed bundle. It checks if the update is signed since there are apps that are not code signed. (This is not a security check, and Sparkle does not intend to fully validate the bundle or identify who signed it in this case). The update validation in Sparkle is complicated, partly due to supporting rotation of (Ed)DSA and Apple code signing. Changes to this path will require careful consideration.

> At any rate, if this problem is fixed then it's possible that Sparkle could already detect SentinelOne's shenanigans without having to introduce other changes.

This does not prove that SentinelOne actually does corrupt the update before Sparkle moves it. And if it does, there could still be a realistic race at the point of time verification is done. I'm also aware there may be some differences between Sparkle 1 and 2 regarding where the update is staged (depending if authorization is used) or when verification is done. The Sparkle 1 path is no longer being iterated on.

**Reply by gnachman** [reply]
Race conditions are unavoidable when trying to solve this problem. I don't believe a 100% perfect solution is possible, but I would happily settle for a 95% solution.

Assuming SentinelOne watches the caches folder that Sparkle extracts to: if an app could opt in to requiring code signing for updates via an Info.plist key then some amount of SentinelOne's damage would be mitigated. The actual amount won't be known until the code is out in the world. Introducing a configurable delay would let developers trade off between performance and broken installs.

Since I run a fork of Sparkle anyway, I've decided to take a more aggressive approach which I don't think is appropriate to upstream but I'll share it in case anyone else is interested: https://github.com/gnachman/Sparkle/commit/218c669d99a0d135cdc99ae20c996f1d5e383b0c

It'll be two updates before I know if this approach is successful. I'll report back when I know.

**Reply by zorgiepoo** [reply]
Reflecting more, regardless of this issue, I think modifying Sparkle's policy to do (something like)

```
if (hostIsCodeSigned && !updateIsCodeSigned) {
        SULog(SULogLevelError, @"The original application is code signed but the update does not have a valid signature.");
        return NO;
    }
```

..should be pretty reasonable if Ed(DSA) verification succeeded in general. This shouldn't trip up adhoc signatures too, which should be used these days for those that have no Apple code signing certificate.

**Reply by zorgiepoo** [reply]
> Since I run a fork of Sparkle anyway, I've decided to take a more aggressive approach which I don't think is appropriate to upstream but I'll share it in case anyone else is interested: https://github.com/gnachman/Sparkle/commit/218c669d99a0d135cdc99ae20c996f1d5e383b0c

Some of these changes are overkill, even for a fork. Sleeping for 10 sec on the main thread (or any other thread) is a very bad idea (particularly also for users that have that SentinelOne set up in a way that allow-listed your app). Since your fork is based on Sparkle 1,  I believe Sparkle 1 validates the update when the user clicks install and relaunch, which for it is "as late as possible" within your app but something could theoretically go wrong while installing the app still or I guess if the user instantly clicks the button (this is also partly a reason why installing updates is so slow in Sparkle 1 and can cause a short hang on install/quit for validating the update if you are observant). Better to check if sentientel one is present after checking the update is not code signed, too.

edit: I would also caution you against hardcoding "SentinelOne" strings or looking if their processes are alive in your program. That sounds like a great potential way to be even more flagged than you currently are.

**Reply by zorgiepoo** [reply]
If a currently impacted user can run (if they're allowed to):

> sudo fs_usage -w | grep $YOUR_APP_NAME > ~/Desktop/logs.txt

then, download and install an update in your app, have SentinelOne tamper the app,

and then Control-C that process, I believe we get diagnostics for at least which part of the update process SentinelOne is interfering and replacing the binary on their machine.
